frontend/exporter/src/id_table.rs

```diff
@@ -0,0 +1,375 @@
+                            })
+                        });
+                        let table = std::mem::replace(
+                            &mut *DESERIALIZATION_STATE.lock().unwrap(),
+                            previous,
+                        );
+                        let (value_key, value) = value_result?;
+                        if !matches!(value_key, Field::Value) {
+                            Err(de::Error::custom(&format!(
+                                "Expected field `value`, found {:#?}",
+                                value_key
+                            )))?
+                        }
+                        if let Some(field) = map.next_key()? {
+                            Err(de::Error::unknown_field(field, &["nothing left"]))?
+                        }
+                        Ok((table, value))
+                    },
+                )?;
+                Ok(Self::Value { table, value })
+            }
+        }
+
+        const FIELDS: &[&str] = &["table", "value"];
+        let r = deserializer.deserialize_struct("WithTable", FIELDS, WithTableVisitor(PhantomData));
+        r
+    }
+}
+
+/// Defines representations for various types when serializing or/and
+/// deserializing via serde
+mod serde_repr {
+    use super::*;
+
+    #[derive(Serialize, Deserialize, JsonSchema, Debug)]
+    pub(super) struct NodeRepr<T> {
+        cache_id: Id,
+        value: Option<Arc<T>>,
+    }
+
+    #[derive(Serialize)]
+    pub(super) struct Pair(Id, Value);
+    pub(super) type SortedIdValuePairs = Vec<Pair>;
+
+    impl<T: SupportedType> Into<NodeRepr<T>> for Node<T> {
+        fn into(self) -> NodeRepr<T> {
+            let value = if serialize_use_id() {
+                None
+            } else {
+                Some(self.value.clone())
+            };
+            let cache_id = self.id;
+            NodeRepr { value, cache_id }
+        }
+    }
+
+    impl<T: 'static + SupportedType> TryFrom<NodeRepr<T>> for Node<T> {
+        type Error = serde::de::value::Error;
+
+        fn try_from(cached: NodeRepr<T>) -> Result<Self, Self::Error> {
+            use serde::de::Error;
+            let map = DESERIALIZATION_STATE.lock().unwrap();
+            let id = cached.cache_id;
+            let kind = if let Some(kind) = cached.value {
+                kind
+            } else {
+                map.get(&id)
+                    .ok_or_else(|| {
+                        Self::Error::custom(&format!(
+                            "Stateful deserialization failed for id {:?}: not found in cache",
+                            id
+                        ))
+                    })?
+                    .ok_or_else(|| {
+                        Self::Error::custom(&format!(
+                            "Stateful deserialization failed for id {:?}: wrong type",
+                            id
+                        ))
+                    })?
+            };
+            Ok(Self { value: kind, id })
+        }
+    }
```

Comment:
You mean calling `DeserializeSeed::*` from a `Deserialize` method using the global state? Or something fancier?

Reply:
No, I was hoping to get rid of global state entirely. I was thinking something like
```rust
impl<T> DeserializeSeed for &mut Option<Table> {
    type Value = WithTable<T>;
    ...
}
```
that would set the table in the `Self` mutable reference and use that throughout. Not entirely sure how that works to get access to the table when deserializing `Node` tbh

Reply:
So, from what I understand (see https://github.com/serde-rs/serde/issues/881 e.g.), this composes badly with `Deserialize` traits: if a you want to use `DeserializeSeed` on a type `T` that appears on a bigger type `U`, then everything in the middle should have a `DeserializeSeed`.
Obviously, that is pretty terrible in our case, given the number of types.

Reply:
I was just reading that same issue ^^ Indeed I am now convinced this would require support from the derive macro but [that's not happening](https://github.com/serde-rs/serde/issues/881#issuecomment-1627865581).

Reply:
^^'
Yeah, exactly, so we have to keep this global state thing sadly :'( :vomiting_face: